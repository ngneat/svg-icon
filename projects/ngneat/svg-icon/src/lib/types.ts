import { InjectionToken } from '@angular/core';

export interface SvgIconType {
  name: string;
  data: string;
}

export interface SVG_CONFIG {
  icons?: SvgIconType | SvgIconType[];
  color?: string;
  defaultSize?: keyof SVG_CONFIG['sizes'];
  missingIconFallback?: SvgIconType;
  sizes: {
    xs?: string;
    sm?: string;
    md?: string;
    lg?: string;
    xl?: string;
    xxl?: string;
  } & Record<string, string>;
}

export const SVG_ICONS_CONFIG = new InjectionToken<SVG_CONFIG>('SVG_ICONS_CONFIG');
export const SVG_ICONS = new InjectionToken<SVG_CONFIG['icons']>('SVG_ICONS');
export const SVG_MISSING_ICON_FALLBACK = new InjectionToken<SVG_CONFIG['missingIconFallback']>('SVG_MISSING_ICON_FALLBACK');
