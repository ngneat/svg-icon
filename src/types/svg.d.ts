export {};
declare global {
    interface SvgIcons {
        icons: 'dashboard' | 'home' | 'user' | 'dashboard-two' | 'location-two' | 'location' | 'missing' | 'settings';
    }
}
