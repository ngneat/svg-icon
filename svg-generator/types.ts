export interface Config {
  srcPath: string;
  outputPath: string;
  svgoConfig: { plugins: any[] };
  prefix?: string;
  postfix?: string;
  rootBarrelFile?: boolean;
  rootBarrelFileName?: string;
  typesPath?: string;
}

export const defaults: Config = {
  prefix: '',
  postfix: 'Icon',
  svgoConfig: { plugins: [] },
  srcPath: '',
  outputPath: '',
  rootBarrelFile: false,
  rootBarrelFileName: 'index',
  typesPath: './src',
};
