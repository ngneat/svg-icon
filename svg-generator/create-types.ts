import {
  createPrinter,
  NewLineKind,
  createSourceFile,
  ScriptTarget,
  ScriptKind,
  factory,
  NodeFlags,
  SyntaxKind,
} from 'typescript';

const printer = createPrinter({
  newLine: NewLineKind.LineFeed,
});

const sourceFile = createSourceFile('types.ts', '', ScriptTarget.Latest, false, ScriptKind.TS);

export function createTypeFile(names: string[]) {
  const block = [
    factory.createExportDeclaration(undefined, false, factory.createNamedExports([]), undefined, undefined),
    factory.createModuleDeclaration(
      [factory.createModifier(SyntaxKind.DeclareKeyword)],
      factory.createIdentifier('global'),
      factory.createModuleBlock([
        factory.createInterfaceDeclaration(undefined, factory.createIdentifier('SvgIcons'), undefined, undefined, [
          factory.createPropertySignature(
            undefined,
            factory.createIdentifier('icons'),
            undefined,
            factory.createUnionTypeNode(
              names.map((name) => {
                return factory.createLiteralTypeNode(factory.createStringLiteral(name, true));
              })
            )
          ),
        ]),
      ]),
      NodeFlags.ExportContext | NodeFlags.GlobalAugmentation | NodeFlags.ContextFlags
    ),
  ];

  const file = factory.updateSourceFile(sourceFile, block);

  return printer.printFile(file);
}
