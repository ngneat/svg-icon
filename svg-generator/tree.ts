import { optimize } from 'svgo';
import { basename, join } from 'path';
import { createPrinter, createSourceFile, EmitHint, factory, NewLineKind, ScriptKind, ScriptTarget, Statement } from 'typescript';
import { readdirSync, readFileSync } from 'fs-extra';
import { createArrayExport, createImportDeclaration, createStatement } from './ast';
import { GeneratorConfig } from './config';
import camelcase from 'camelcase';

const printer = createPrinter({
  newLine: NewLineKind.LineFeed,
});

const sourceFile = createSourceFile('generator.ts', '', ScriptTarget.Latest, false, ScriptKind.TS);

type VirtualFile = {
  path: string;
  content: string;
  name: string;
  identifierName: string;
}

type Config = Required<Omit<GeneratorConfig, 'srcPath' | 'outputPath'>>;

export const INDEX = `__INDEX__`;

export function createTree(srcPath: string, outputPath: string, config: Config): VirtualFile[] {
  const tree: VirtualFile[] = [];
  const result = readdirSync(srcPath, { withFileTypes: true });
  const plugins = config.svgoConfig?.plugins ?? [];

  for(const file of result) {

    if(file.isDirectory()) {
      const children = createTree(join(srcPath, file.name), join(outputPath, file.name), config);

      const exportDeclarations: Statement[] = [];
      const identifiers: string[] = [];

      for(const { identifierName, name } of children) {
        exportDeclarations.push(createImportDeclaration({ identifierName, iconName: name }));
        identifiers.push(identifierName);
      }

      exportDeclarations.push(createArrayExport(file.name, identifiers));
      const barrelFile = factory.updateSourceFile(sourceFile, exportDeclarations);

      tree.push(...children, {
        path: join(outputPath, file.name, `index.ts`),
        content: printer.printFile(barrelFile),
        identifierName: 'index',
        name: INDEX
      });

    } else {
      if(file.name.endsWith('.svg')) {
        const iconName = basename(file.name, '.svg');
        const path = join(outputPath, file.name).replace('.svg', '.ts');
        const identifierName = resolveIdentifierName(iconName, config);
        const svgPath = join(srcPath, file.name);
        const svgContent = readFileSync(svgPath).toString();

        const statement = createStatement({
          svgContent: optimize(svgContent, { plugins, path: svgPath }).data,
          iconName,
          identifierName,
        });

        tree.push({
          path,
          content: printer.printNode(EmitHint.Unspecified, statement, sourceFile),
          name: iconName,
          identifierName
        });
      }

    }
  }

  return tree;
}

const invalidVariableCharsRegex = /[^a-zA-Z0-9_$]/g;
const startWithDigitRegex = /^[0-9]/;
const lettersRegex = /[a-zA-Z]/;

export function resolveIdentifierName(iconName: string, config: Config) {
  return camelcase([config.prefix, iconName, config.postfix])
    // Replace invalid characters with $ sign
    .replace(invalidVariableCharsRegex, config.invalidCharReplacer)
    // Ensure the first character is not a digit
    .replace(startWithDigitRegex, config.invalidCharReplacer)
    // Ensure the first letter is lowercase
    .replace(lettersRegex, (match) => match.toLowerCase());
}
