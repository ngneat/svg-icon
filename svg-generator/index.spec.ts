import mock from 'mock-fs';
import { createTypeFile } from './create-types';
import { createTree } from './tree';
import { defaults } from './types';

const srcPath = `src/assets/svg`;

describe('createTree', () => {
  it('should create the correct tree', () => {
    mock({
      [`${srcPath}/one.svg`]: '<svg>one</svg>',
      [`${srcPath}/two.svg`]: '<svg>two</svg>',
      [`${srcPath}/group`]: {
        'child.svg': `<svg>child</svg>`,
        'child2.svg': `<svg>child 2</svg>`,
      },
      [`${srcPath}/group-two`]: {
        'child-one.svg': `<svg>child</svg>`,
        'child-two.svg': `<svg>child 2</svg>`,
      },
    });

    const result = createTree(srcPath, `src/app/svg`, defaults);

    mock.restore();

    expect(result).toMatchSnapshot();
  });

  it('should create the type file', () => {
    expect(createTypeFile(['foo', 'bar', 'foo-bar'])).toMatchSnapshot();
  });
});
