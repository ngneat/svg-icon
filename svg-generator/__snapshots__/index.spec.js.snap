// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTree should create the correct tree 1`] = `
[
  {
    "content": "export const childIcon = {
    data: \`<svg>child</svg>\`,
    name: 'child' as const
};",
    "identifierName": "childIcon",
    "name": "child",
    "path": "src/app/svg/group/child.ts",
  },
  {
    "content": "export const child2Icon = {
    data: \`<svg>child 2</svg>\`,
    name: 'child-2' as const
};",
    "identifierName": "child2Icon",
    "name": "child2",
    "path": "src/app/svg/group/child2.ts",
  },
  {
    "content": "import { childIcon } from './child';
import { child2Icon } from './child2';
export const groupIcons = [childIcon, child2Icon];
",
    "identifierName": "index",
    "name": "__INDEX__",
    "path": "src/app/svg/group/index.ts",
  },
  {
    "content": "export const childOneIcon = {
    data: \`<svg>child</svg>\`,
    name: 'child-one' as const
};",
    "identifierName": "childOneIcon",
    "name": "child-one",
    "path": "src/app/svg/group-two/child-one.ts",
  },
  {
    "content": "export const childTwoIcon = {
    data: \`<svg>child 2</svg>\`,
    name: 'child-two' as const
};",
    "identifierName": "childTwoIcon",
    "name": "child-two",
    "path": "src/app/svg/group-two/child-two.ts",
  },
  {
    "content": "import { childOneIcon } from './child-one';
import { childTwoIcon } from './child-two';
export const groupTwoIcons = [childOneIcon, childTwoIcon];
",
    "identifierName": "index",
    "name": "__INDEX__",
    "path": "src/app/svg/group-two/index.ts",
  },
  {
    "content": "export const oneIcon = {
    data: \`<svg>one</svg>\`,
    name: 'one' as const
};",
    "identifierName": "oneIcon",
    "name": "one",
    "path": "src/app/svg/one.ts",
  },
  {
    "content": "export const twoIcon = {
    data: \`<svg>two</svg>\`,
    name: 'two' as const
};",
    "identifierName": "twoIcon",
    "name": "two",
    "path": "src/app/svg/two.ts",
  },
]
`;
